src/components/v5/common/CompletedAction/partials/PaymentBuilder/partials/PaymentBuilderWidget/utils.ts

```diff
@@ -52,6 +52,138 @@ export const isExpenditureFullyFunded = (expenditure?: Expenditure | null) => {
   });
 };
 
+export const getFundingItemIndex = (
+  fundingActions: ExpenditureActionFragment[][] | undefined,
+) => {
+  if (!fundingActions || fundingActions.length === 0) {
+    return 0;
+  }
+
+  const lastFundingItem = fundingActions[fundingActions.length - 1];
+  const lastItem = lastFundingItem[lastFundingItem.length - 1];
+
+  if (
+    !lastItem.motionData ||
+    lastItem.motionData.motionStateHistory.hasPassed
+  ) {
+    return fundingActions.length;
+  }
+
+  return fundingActions.length - 1;
+};
+
+export const segregateFundingActions = (
+  expenditure: Expenditure | null | undefined,
+) => {
+  if (!expenditure) {
+    return undefined;
+  }
+
+  const result: ExpenditureAction[][] = [];
+
+  const { fundingActions } = expenditure;
+
+  if (!fundingActions?.items || fundingActions.items.length === 0) {
+    return result;
+  }
+
+  const seenActions = new Set();
+
+  const sortedFundingActions = [...fundingActions.items].sort((a, b) => {
+    if (a?.createdAt && b?.createdAt) {
+      return Date.parse(a.createdAt) - Date.parse(b.createdAt);
+    }
+    return 0;
+  });
+
+  sortedFundingActions.forEach((action, index) => {
+    if (!action) {
+      return;
+    }
+
+    const prevAction = sortedFundingActions[index - 1];
+
+    if (prevAction && prevAction.motionData) {
+      const { hasFailed, hasFailedNotFinalizable, hasPassed } =
+        prevAction.motionData.motionStateHistory;
+
+      if ((hasFailed || hasFailedNotFinalizable) && !hasPassed) {
+        result[result.length - 1].push(action);
+        return;
+      }
+    }
+
+    const { createdAt, transactionHash } = action;
+    if (!createdAt || !transactionHash) return;
+
+    const fundingActionTime = new Date(createdAt);
+    const actionKey = `${createdAt}-${transactionHash}`;
```

Comment:
transactionHash, for now, is a unique identifier of an action so you can just use that